source/common/common/trie.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+namespace {
+// Because vector is efficient when removing elements from the back but paths
+// generally have the most significant components at the front, it's useful to
+// reverse the components vector before processing
+template <class T> inline void reverseComponentVector(std::vector<T>& vec) {
+  std::reverse(vec.begin(), vec.end());
+}
+} // namespace
+
+/**
+ * This is a general purpose trie. Keys and Values can be arbitrary classes
+ * provided they can be hashed. It supports both partial matches and exact
+ * matches.
```

Comment:
Please add a comment explaining the Value must be tokenizable by Tokenizer and that the trie uses these tokens for its path structure.